### PR TITLE
Add sample method in LightCurveTableModel and PhaseCurveTableModel

### PR DESCRIPTION
**Pull request by fabiopintore**
This is second PR listed in the PIG 9 and aims at implementing a sampling method of the time into `gammapy.time.models.LightCurveTableModel` and `gammapy.time.models.PhaseCurveTableModel`. 
Tests have been added.

### CODE REVIEW

File: gammapy/modeling/models/time.py [hunk 1]
```diff
@@ -246,3 +316,71 @@ def mean_norm_in_time_interval(self, time_min, time_max):
...
+    def ontime(self, t_min, t_max):
+        """On time (`~astropy.units.Quantity`)"""
+
+        t_min = Time(t_min)
+        t_max = Time(t_max)
+
+        ontime = t_max - t_min
+        return u.Quantity(ontime.sec, "s")
+
+    def _get_time_meta(self, t_min, t_max):
```
Comment (adonath) [inline gammapy/modeling/models/time.py]: As far as I can tell this method is unused. Please remove.

File: gammapy/modeling/models/time.py [hunk 2]
```diff
@@ -123,6 +125,74 @@ def evaluate_norm_at_phase(self, phase):
         fp = self.table["NORM"]
         return np.interp(x=phase, xp=xp, fp=fp, period=1)
 
+    #    @property
+    def ontime(self, t_min, t_max):
```
Comment (adonath) [inline gammapy/modeling/models/time.py]: I think there is no real need to have this in a separate method anymore please just move the code to `sample_time()` and remove the separate method.

File: gammapy/modeling/models/time.py [hunk 3]
```diff
@@ -123,6 +125,74 @@ def evaluate_norm_at_phase(self, phase):
...
+        t_max = Time(t_max)
+        n_events = n_events
+        t_delta = u.Quantity(t_delta)
+        random_state = get_random_state(random_state)
+
+        t_stop = self.ontime(t_min=t_min, t_max=t_max).to_value(time_unit)
+
+        # TODO: the separate time unit handling is unfortunate, but the quantity support for np.arange and np.interp
+        #  is still incomplete, refactor once we change to recent numpy and astropy versions
+        if self.table is not None:
```
Comment (adonath) [inline gammapy/modeling/models/time.py]: Remove this logic of `if self.table is not None`. The case of a uniform time distribution we will handle outside. I think it's not useful to have it attached to the model.

File: gammapy/modeling/models/time.py [hunk 4]
```diff
@@ -246,3 +316,71 @@ def mean_norm_in_time_interval(self, time_min, time_max):
...
+        """
+        time_unit = u.second
+
+        t_min = Time(t_min)
+        t_max = Time(t_max)
+        n_events = n_events
+        t_delta = u.Quantity(t_delta)
+        random_state = get_random_state(random_state)
+
+        t_stop = self.ontime(t_min=t_min, t_max=t_max).to_value(time_unit)
```
Comment (adonath) [inline gammapy/modeling/models/time.py]: See my comment above...

File: gammapy/modeling/models/time.py [hunk 5]
```diff
@@ -246,3 +316,71 @@ def mean_norm_in_time_interval(self, time_min, time_max):
...
+
+        # TODO: the separate time unit handling is unfortunate, but the quantity support for np.arange and np.interp
+        #  is still incomplete, refactor once we change to recent numpy and astropy versions
+        if self.table is not None:
```
Comment (adonath) [inline gammapy/modeling/models/time.py]: See my comment above...

File: gammapy/modeling/models/time.py [hunk 6]
```diff
@@ -123,6 +125,74 @@ def evaluate_norm_at_phase(self, phase):
...
+    def ontime(self, t_min, t_max):
+        """On time (`~astropy.units.Quantity`)"""
+
+        t_min = Time(t_min)
+        t_max = Time(t_max)
+
+        ontime = t_max - t_min
+        return u.Quantity(ontime.sec, "s")
+
+    def _get_time_meta(self, t_min, t_max):
```
Comment (adonath) [inline gammapy/modeling/models/time.py]: As far as I can tell this method is unused. Please remove.

File: gammapy/modeling/models/time.py [hunk 7]
```diff
@@ -246,3 +316,71 @@ def mean_norm_in_time_interval(self, time_min, time_max):
         dt = time_max - time_min
         integral = self._interpolator.integral(time_min, time_max)
         return integral / dt
+
+    #    @property
+    def ontime(self, t_min, t_max):
```
Comment (adonath) [inline gammapy/modeling/models/time.py]: See my comment above...

File: gammapy/modeling/models/tests/test_time.py [hunk 8]
```diff
@@ -58,3 +61,59 @@ def test_light_curve_evaluate_norm_at_time(light_curve):
...
+    temporal_model = LightCurveTableModel(table)
+
+    t_min = "2010-01-01T00:00:00"
+    t_max = "2010-01-01T08:00:00"
+
+    sampler = temporal_model.sample_time(
+        n_events=2, t_min=t_min, t_max=t_max, random_state=0, t_delta="10 min"
+    )
+
+    sampler_no_lc = LightCurveTableModel(None).sample_time(
```
Comment (adonath) [inline gammapy/modeling/models/tests/test_time.py]: It think for this case we should just introduce a `TemporalConstant` time model. Please just remove the test case...

File: gammapy/modeling/models/time.py (hunk 1 above)
Comment (fabiopintore) [reply gammapy/modeling/models/time.py]: Done

File: gammapy/modeling/models/time.py (hunk 2 above)
Comment (fabiopintore) [reply gammapy/modeling/models/time.py]: Done

File: gammapy/modeling/models/time.py (hunk 4 above)
Comment (fabiopintore) [reply gammapy/modeling/models/time.py]: Done

File: gammapy/modeling/models/time.py (hunk 6 above)
Comment (fabiopintore) [reply gammapy/modeling/models/time.py]: Done

File: gammapy/modeling/models/time.py (hunk 7 above)
Comment (fabiopintore) [reply gammapy/modeling/models/time.py]: Done

File: gammapy/modeling/models/time.py [hunk 9]
```diff
@@ -123,6 +125,55 @@ def evaluate_norm_at_phase(self, phase):
...
+        Returns
+        -------
+        time : `~astropy.units.Quantity`
+            Array with times of the sampled events.
+        """
+        time_unit = u.second
+
+        t_min = Time(t_min)
+        t_max = Time(t_max)
+        n_events = n_events
```
Comment (adonath) [inline gammapy/modeling/models/time.py]: This statement is without effect, just remove...

File: gammapy/modeling/models/time.py (hunk 3 above)
Comment (adonath) [reply gammapy/modeling/models/time.py]: Please still address this comment...

File: gammapy/modeling/models/time.py [hunk 10]
```diff
@@ -246,3 +297,52 @@ def mean_norm_in_time_interval(self, time_min, time_max):
...
```
Comment (adonath) [inline gammapy/modeling/models/time.py]: This statement is without effect, please remove...

File: gammapy/modeling/models/time.py [hunk 11]
```diff
@@ -123,6 +125,55 @@ def evaluate_norm_at_phase(self, phase):
...
+
+            pdf = self.evaluate_norm_at_time(t)
+
+            sampler = InverseCDFSampler(pdf=pdf, random_state=random_state)
+            time_pix = sampler.sample(n_events)[0]
+            time = np.interp(time_pix, np.arange(len(t)), t) * time_unit
+        else:
+            time = random_state.uniform(high=t_stop, size=n_events) * time_unit
+
+        return time
```
Comment (adonath) [inline gammapy/modeling/models/time.py]: Maybe we should directly return absolute time i.e. `t_min + time`?

File: gammapy/modeling/models/time.py (hunk 3 above)
Comment (fabiopintore) [reply gammapy/modeling/models/time.py]: If I remove it, then the user is forced to always pass a lightcurve. Is what we really want?

File: gammapy/modeling/models/time.py (hunk 5 above)
Comment (fabiopintore) [reply gammapy/modeling/models/time.py]: See above

File: gammapy/modeling/models/tests/test_time.py (hunk 8 above)
Comment (fabiopintore) [reply gammapy/modeling/models/tests/test_time.py]: Done.

File: gammapy/modeling/models/time.py (hunk 9 above)
Comment (fabiopintore) [reply gammapy/modeling/models/time.py]: Done.

File: gammapy/modeling/models/time.py (hunk 10 above)
Comment (fabiopintore) [reply gammapy/modeling/models/time.py]: Done.

File: gammapy/modeling/models/time.py (hunk 11 above)
Comment (fabiopintore) [reply gammapy/modeling/models/time.py]: Done.

**Comment by adonath** [comment]
Thanks @fabiopintore!

**Pull request merged**